s3/s3_transport/include/s3_transport_callbacks.hpp

```diff
@@ -0,0 +1,814 @@
+                                             void *callback_data );
+
+            libs3_types::status on_response_properties (const libs3_types::response_properties *properties,
+                                                        void *callback_data);
+
+            void on_response_complete (libs3_types::status status,
+                                       const libs3_types::error_details *error,
+                                       void *callback_data);
+        } // end namespace initialization_callback
+
+
+        /* Uploading the multipart completion XML from our buffer */
+        namespace commit_callback
+        {
+            int on_response (int buffer_size,
+                          libs3_types::buffer_type buffer,
+                          void *callback_data);
+
+            libs3_types::status on_response_properties (const libs3_types::response_properties *properties,
+                                                        void *callback_data);
+
+            void on_response_completion (libs3_types::status status,
+                                         const libs3_types::error_details *error,
+                                         void *callback_data);
+
+        } // end namespace commit_callback
+
+
+        template <typename buffer_type>
+        class callback_for_write_to_s3_base
+        {
+
+            public:
+
+                callback_for_write_to_s3_base(libs3_types::bucket_context& _saved_bucket_context,
+                                              upload_manager& _manager)
+                    : enable_md5{false}
+                    , thread_identifier{0}
+                    , shared_memory_timeout_in_seconds{constants::DEFAULT_SHARED_MEMORY_TIMEOUT_IN_SECONDS}
+                    , object_key{}
+                    , shmem_key{}
+                    , sequence{0}
+                    , offset{0}
+                    , content_length{0}
+                    , saved_bucket_context{_saved_bucket_context}
+                    , manager{_manager}
+                    , bytes_written{0}
+                    , callback_counter{0}
+                {}
+
+
+                virtual int callback_implementation(int libs3_buffer_size,
+                                                    libs3_types::char_type *libs3_buffer) = 0;
+
+                static int invoke_callback(int libs3_buffer_size,
+                                           libs3_types::char_type *libs3_buffer,
+                                           void *callback_data)
+                {
+
+                    using named_shared_memory_object =
+                        irods::experimental::interprocess::shared_memory::named_shared_memory_object
+                        <shared_data::multipart_shared_data>;
+
+                    callback_for_write_to_s3_base *data =
+                        (callback_for_write_to_s3_base*)callback_data;
+
+                    // just touch shmem so we know we are active
```

Comment:
i'm seeing this 3-4x - please either pull this out into a call... or create an issue for later

Reply:
Not sure I understand.

There is a timeout in the shared memory.  This is to recover from a failure.  The "touch" is done so that we don't accidentally time out during processing.

It is overkill right now as it is touched far more often than the timeout parameter

Reply:
so they're not quite the same - let's do it later.